### ISSUE TITLE
[EC2 driver] Docker Machine exposed to the world?

### ISSUE THREAD

**Issue by cmpitg** [docker/machine#2543]
Hi,

CC'ing @adhorn

I have some questions related to Docker Machine Amazon EC2 driver:

* Why is the security group [patched](https://github.com/docker/machine/blob/95bd94da625615be9ce9841a2518b8c377314e69/drivers/amazonec2/amazonec2.go#L655)? Would it be better to create a new one just for Docker Machine and Docker Swarm instead of patching the current one? I have other instances depending on this security group as well. If the security group is modified, should there be a **bold** notification?

* Why exposing the instance to [the](https://github.com/docker/machine/blob/95bd94da625615be9ce9841a2518b8c377314e69/drivers/amazonec2/amazonec2.go#L26) [world](https://github.com/docker/machine/blob/95bd94da625615be9ce9841a2518b8c377314e69/drivers/amazonec2/amazonec2.go#L699)? This a **serious serious** security issue. I understand why Docker/Swarm port needs to be opened, but should it be documented so that users are aware and could open them manually? The current behavior is not at all sane.

Ha-Duong

**Comment by nathanleclaire** [comment]
> Why is the security group patched? Would it be better to create a new one just for Docker Machine and Docker Swarm instead of patching the current one? I have other instances depending on this security group as well. If the security group is modified, should there be a bold notification?

Yes, I agree this design is not ideal, and would prefer to just create a Docker-Swarm-master enabled security group.

> Why exposing the instance to the world? This a serious serious security issue. I understand why Docker/Swarm port needs to be opened, but should it be documented so that users are aware and could open them manually? The current behavior is not at all sane.

What would you consider the best approach?  The current behavior, while not ideal, still protects the exposed API ports through TLS.  Requiring users to also modify security group rules increases the overhead from `docker-machine create` to "Running a Docker command remotely and it works".

**Comment by cmpitg** [comment]
> Yes, I agree this design is not ideal, and would prefer to just create a Docker-Swarm-master enabled security group.

+1

Also, this should be done with user's consensus.

Regarding the exposure of ports, how about adding some flags? For example:

```
# Allowing inbound traffic from a list of security groups
--inbound-from-security-groups sg1,sg2

# Allowing inbound traffic from subnet(s)
--vpc-id id --inbound-from-cidr "x.x.x.x/x,x.x.x.x/x"
```

Ha-Duong

**Comment by cmpitg** [comment]
Any updates on this matter?

Ha-Duong

**Comment by nathanleclaire** [comment]
@cmpitg We're taking it into considering for future improvements and/or potential re-write of the EC2 driver.

**Comment by cmpitg** [comment]
@nathanleclaire  Thank you.  It's good to know.  Please keep this issue posted.

Ha-Duong

**Comment by cyphrsonic** [comment]
Are there any plans for adding this support for custom ip ranges in the future? This seems like a serious security issue. Even if the endpoint is TLS secured, we'd prefer to have the traffic locked down to a specific IP range.
